File: Data Structures & Algorithms/buy-and-sell-crypto/submission-2.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) < 2:
            return 0
 
        l = 0
        r = 1
        res = 0
        while (r < len(prices)):
            if prices[l] < prices[r]:
                profit = prices[r] - prices[l]
                res = max(res, profit)
            else:
                l = r
            r += 1
        return res


"""
Input: prices = [10,1,5,6,7,1]

Output: 6

l = 0
r = 1
res = 6
min(l)= 10
max(r) = 1

to-do
- if l index is on index r, r increments 
- l or r keeps indexing depends on which has more profit, else would be indexing r
"""

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-2.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        unordered_map<char, int> map;
        int l = 0, res = 0, maxf = 0;
        for (int i = 0; i < s.size(); i++) {
            map[s[i]]++;
            maxf = max(maxf, map[s[i]]);
            while ((i-l+1)-maxf > k) {
                map[s[l]]--;
                l++;
            }
            res = max(res, i-l+1);
        }

        return res;
    }
};
